File: Examples/Java/Onboard and MXP/src/main/java/frc/robot/Robot.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2017-2018 FIRST. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package frc.robot;

import edu.wpi.first.wpilibj.IterativeRobot;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

import com.revrobotics.*;
import com.revrobotics.Rev2mDistanceSensor.Port;

public class Robot extends IterativeRobot {
  private Rev2mDistanceSensor distOnboard; 
  private Rev2mDistanceSensor distMXP;

  @Override
  public void robotInit() {
    /**
     * Rev 2m distance sensor can be initialized with the Onboard I2C port
     * or the MXP port. Both can run simultaneously.
     */
    distOnboard = new Rev2mDistanceSensor(Port.kOnboard);
    distMXP = new Rev2mDistanceSensor(Port.kMXP);
  }

  @Override
  public void robotPeriodic() {}

  @Override
  public void autonomousInit() {}

  @Override
  public void autonomousPeriodic() {}

  @Override
  public void teleopInit() {
    /**
     * Before measurements can be read from the sensor, setAutomaticMode(true)
     * must be called. This starts a background thread which will periodically
     * poll all enabled sensors and store their measured range.
     */
    distOnboard.setAutomaticMode(true);
  }

  @Override
  public void teleopPeriodic() {
    /**
     * Range returned from the distance sensor is valid if isRangeValid()
     * returns true.
     */
    if(distOnboard.isRangeValid()) {
      SmartDashboard.putNumber("Range Onboard", distOnboard.getRange());
      SmartDashboard.putNumber("Timestamp Onboard", distOnboard.getTimestamp());
    }

    if(distMXP.isRangeValid()) {
      SmartDashboard.putNumber("Range MXP", distMXP.getRange());
      SmartDashboard.putNumber("Timestamp MXP", distMXP.getTimestamp());
    }
  }

  @Override
  public void testPeriodic() {}

  @Override
  public void disabledInit() {
    /**
     * The background thread may be stopped by calling setAutomaticMode(false).
     * This will command any active sensors to terminate current measurements
     * and the thread will stop.
     */
    distOnboard.setAutomaticMode(false);
  }
}

File: Examples/C++/Read Range/src/main/cpp/Robot.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2017-2018 FIRST. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include "Robot.h"

#include <iostream>

#include <frc/smartdashboard/SmartDashboard.h>

#include "rev/Rev2mDistanceSensor.h"

void Robot::RobotInit() {
  // initialize distance sensor on OnBoard port, set units to inches
  distSensor = new rev::Rev2mDistanceSensor{rev::Rev2mDistanceSensor::Port::kOnboard, 
                                            rev::Rev2mDistanceSensor::DistanceUnit::kInches};
}

void Robot::RobotPeriodic() {}

void Robot::AutonomousInit() {}

void Robot::AutonomousPeriodic() {}

void Robot::TeleopInit() {
  /**
   * Call SetAutomaticMode(true) to begin background autonomous
   * thread. Will periodically ping sensor for new data 
   */
  distSensor->SetAutomaticMode(true);

  /**
   * By default, sensor is enabled. However, SetEnabled(false) is called
   * in DisabledInit() below, so we need to reenable every time we go into
   * teleop mode
   */
  distSensor->SetEnabled(true);
}

void Robot::TeleopPeriodic() {
  /**
   * The current measurement is considered valid if IsRangeValid()
   * returns true.
   */
  bool isValid = distSensor->IsRangeValid();

  frc::SmartDashboard::PutBoolean("Data Valid", isValid);

  if(isValid) {
    /**
     * The current measured range is returned from GetRange(). By default
     * this range is returned in inches.
     */
    frc::SmartDashboard::PutNumber("Distance (in)", distSensor->GetRange());

    /**
     * The timestamp of the last valid measurement (measured in seconds since 
     * the program started), is returned by GetTimestamp().
     */
    frc::SmartDashboard::PutNumber("Timestamp", distSensor->GetTimestamp());
  }
  else {
    frc::SmartDashboard::PutNumber("Distance (in)", -1);
  }
}

void Robot::TestPeriodic() {}

void Robot::DisabledInit() {
  /**
   * The background round robin thread can be stopped by calling 
   * SetAutomaticMode(false)
   */
  distSensor->SetAutomaticMode(false);

  /**
   * In order to completely kill the background thread, all current
   * sensors need to be disabled by calling SetEnabled(false)
   */
  distSensor->SetEnabled(false);
}

#ifndef RUNNING_FRC_TESTS
int main() { return frc::StartRobot<Robot>(); }
#endif
